Author: Marvin Humphrey <marvin@rectangular.com>
Date: Fri, 7 Oct 2011 19:55:44 -0700
Bug: http://sourceforge.net/tracker/?func=detail&aid=3420484&group_id=55736&atid=478072
Bug: http://sourceforge.net/tracker/?func=detail&aid=3421072&group_id=55736&atid=478072
Description: Improve UTF-8 processing.
 Pd's UTF-8 handling library assumes NUL-termination of strings and often
 malfunctions in its absence, but Pd uses strings without NUL-termination.  To
 guard agains invalid reads, this patch changes u8_offset, u8_charnum and
 u8_inc so that they function properly without NUL-termination.
 .
 An additional benefit of this patch is that all three functions are sped up.
 .
 No problems were found in u8_dec and it could not be sped up, so it has been
 left alone.  Some other functions may or may not have problems (e.g.
 u8_inc_ptr), but do not seem to be in use.

--- puredata.orig/src/s_utf8.c
+++ puredata/src/s_utf8.c
@@ -17,138 +17,9 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
-#include <stdarg.h>
-#ifdef WIN32
-#include <malloc.h>
-#else
-#include <alloca.h>
-#endif
 
 #include "s_utf8.h"
 
-static const u_int32_t offsetsFromUTF8[6] = {
-    0x00000000UL, 0x00003080UL, 0x000E2080UL,
-    0x03C82080UL, 0xFA082080UL, 0x82082080UL
-};
-
-static const char trailingBytesForUTF8[256] = {
-    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
-    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
-    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
-    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
-    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
-    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
-    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, 1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
-    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2, 3,3,3,3,3,3,3,3,4,4,4,4,5,5,5,5
-};
-
-
-/* returns length of next utf-8 sequence */
-int u8_seqlen(char *s)
-{
-    return trailingBytesForUTF8[(unsigned int)(unsigned char)s[0]] + 1;
-}
-
-/* conversions without error checking
-   only works for valid UTF-8, i.e. no 5- or 6-byte sequences
-   srcsz = source size in bytes, or -1 if 0-terminated
-   sz = dest size in # of wide characters
-
-   returns # characters converted
-   dest will always be L'\0'-terminated, even if there isn't enough room
-   for all the characters.
-   if sz = srcsz+1 (i.e. 4*srcsz+4 bytes), there will always be enough space.
-*/
-int u8_toucs(u_int32_t *dest, int sz, char *src, int srcsz)
-{
-    u_int32_t ch;
-    char *src_end = src + srcsz;
-    int nb;
-    int i=0;
-
-    while (i < sz-1) {
-        nb = trailingBytesForUTF8[(unsigned char)*src];
-        if (srcsz == -1) {
-            if (*src == 0)
-                goto done_toucs;
-        }
-        else {
-            if (src + nb >= src_end)
-                goto done_toucs;
-        }
-        ch = 0;
-        switch (nb) {
-            /* these fall through deliberately */
-#if UTF8_SUPPORT_FULL_UCS4
-        case 5: ch += (unsigned char)*src++; ch <<= 6;
-        case 4: ch += (unsigned char)*src++; ch <<= 6;
-#endif
-        case 3: ch += (unsigned char)*src++; ch <<= 6;
-        case 2: ch += (unsigned char)*src++; ch <<= 6;
-        case 1: ch += (unsigned char)*src++; ch <<= 6;
-        case 0: ch += (unsigned char)*src++;
-        }
-        ch -= offsetsFromUTF8[nb];
-        dest[i++] = ch;
-    }
- done_toucs:
-    dest[i] = 0;
-    return i;
-}
-
-/* srcsz = number of source characters, or -1 if 0-terminated
-   sz = size of dest buffer in bytes
-
-   returns # characters converted
-   dest will only be '\0'-terminated if there is enough space. this is
-   for consistency; imagine there are 2 bytes of space left, but the next
-   character requires 3 bytes. in this case we could NUL-terminate, but in
-   general we can't when there's insufficient space. therefore this function
-   only NUL-terminates if all the characters fit, and there's space for
-   the NUL as well.
-   the destination string will never be bigger than the source string.
-*/
-int u8_toutf8(char *dest, int sz, u_int32_t *src, int srcsz)
-{
-    u_int32_t ch;
-    int i = 0;
-    char *dest_end = dest + sz;
-
-    while (srcsz<0 ? src[i]!=0 : i < srcsz) {
-        ch = src[i];
-        if (ch < 0x80) {
-            if (dest >= dest_end)
-                return i;
-            *dest++ = (char)ch;
-        }
-        else if (ch < 0x800) {
-            if (dest >= dest_end-1)
-                return i;
-            *dest++ = (ch>>6) | 0xC0;
-            *dest++ = (ch & 0x3F) | 0x80;
-        }
-        else if (ch < 0x10000) {
-            if (dest >= dest_end-2)
-                return i;
-            *dest++ = (ch>>12) | 0xE0;
-            *dest++ = ((ch>>6) & 0x3F) | 0x80;
-            *dest++ = (ch & 0x3F) | 0x80;
-        }
-        else if (ch < 0x110000) {
-            if (dest >= dest_end-3)
-                return i;
-            *dest++ = (ch>>18) | 0xF0;
-            *dest++ = ((ch>>12) & 0x3F) | 0x80;
-            *dest++ = ((ch>>6) & 0x3F) | 0x80;
-            *dest++ = (ch & 0x3F) | 0x80;
-        }
-        i++;
-    }
-    if (dest < dest_end)
-        *dest = '\0';
-    return i;
-}
-
 /* moo: get byte length of character number, or 0 if not supported */
 int u8_wc_nbytes(u_int32_t ch)
 {
@@ -156,11 +27,6 @@
   if (ch < 0x800) return 2;
   if (ch < 0x10000) return 3;
   if (ch < 0x200000) return 4;
-#if UTF8_SUPPORT_FULL_UCS4
-  /*-- moo: support full UCS-4 range? --*/
-  if (ch < 0x4000000) return 5;
-  if (ch < 0x7fffffffUL) return 6;
-#endif
   return 0; /*-- bad input --*/
 }
 
@@ -202,61 +68,63 @@
 /* charnum => byte offset */
 int u8_offset(char *str, int charnum)
 {
-    int offs=0;
+    char *string = str;
 
-    while (charnum > 0 && str[offs]) {
-        (void)(isutf(str[++offs]) || isutf(str[++offs]) ||
-               isutf(str[++offs]) || ++offs);
-        charnum--;
+    while (charnum > 0 && *string != '\0') {
+        if (*string++ & 0x80) {
+            if (!isutf(*string)) {
+                ++string;
+                if (!isutf(*string)) {
+                    ++string;
+                    if (!isutf(*string)) {
+                        ++string;
+                    }
+                }
+            }
+        }
+        --charnum;
     }
-    return offs;
+
+    return (int)(string - str);
 }
 
 /* byte offset => charnum */
 int u8_charnum(char *s, int offset)
 {
-    int charnum = 0, offs=0;
-
-    while (offs < offset && s[offs]) {
-        (void)(isutf(s[++offs]) || isutf(s[++offs]) ||
-               isutf(s[++offs]) || ++offs);
-        charnum++;
+    int charnum = 0;
+    char *string = s;
+    char *const end = string + offset;
+
+    while (string < end && *string != '\0') {
+        if (*string++ & 0x80) {
+            if (!isutf(*string)) {
+                ++string;
+                if (!isutf(*string)) {
+                    ++string;
+                    if (!isutf(*string)) {
+                        ++string;
+                    }
+                }
+            }
+        }
+        ++charnum;
     }
     return charnum;
 }
 
-/* reads the next utf-8 sequence out of a string, updating an index */
-u_int32_t u8_nextchar(char *s, int *i)
-{
-    u_int32_t ch = 0;
-    int sz = 0;
-
-    do {
-        ch <<= 6;
-        ch += (unsigned char)s[(*i)++];
-        sz++;
-    } while (s[*i] && !isutf(s[*i]));
-    ch -= offsetsFromUTF8[sz-1];
-
-    return ch;
-}
-
-/* number of characters */
-int u8_strlen(char *s)
-{
-    int count = 0;
-    int i = 0;
-
-    while (u8_nextchar(s, &i) != 0)
-        count++;
-
-    return count;
-}
-
 void u8_inc(char *s, int *i)
 {
-    (void)(isutf(s[++(*i)]) || isutf(s[++(*i)]) ||
-           isutf(s[++(*i)]) || ++(*i));
+    if (s[(*i)++] & 0x80) {
+        if (!isutf(s[*i])) {
+            ++(*i);
+            if (!isutf(s[*i])) {
+                ++(*i);
+                if (!isutf(s[*i])) {
+                    ++(*i);
+                }
+            }
+        }
+    }
 }
 
 void u8_dec(char *s, int *i)
@@ -265,16 +133,3 @@
            isutf(s[--(*i)]) || --(*i));
 }
 
-/*-- moo --*/
-void u8_inc_ptr(char **sp)
-{
-  (void)(isutf(*(++(*sp))) || isutf(*(++(*sp))) ||
-         isutf(*(++(*sp))) || ++(*sp));
-}
-
-/*-- moo --*/
-void u8_dec_ptr(char **sp)
-{
-  (void)(isutf(*(--(*sp))) || isutf(*(--(*sp))) ||
-         isutf(*(--(*sp))) || --(*sp));
-}
--- puredata.orig/src/s_utf8.h
+++ puredata/src/s_utf8.h
@@ -10,51 +10,26 @@
 # define UCS4 u_int32_t
 #endif
 
-/* UTF8_SUPPORT_FULL_UCS4
- *  define this to support the full potential range of UCS-4 codepoints
- *  (in anticipation of a future UTF-8 standard)
- */
-/*#define UTF8_SUPPORT_FULL_UCS4 1*/
-#undef UTF8_SUPPORT_FULL_UCS4
-
 /* UTF8_MAXBYTES
  *   maximum number of bytes required to represent a single character in UTF-8
  *
  * UTF8_MAXBYTES1 = UTF8_MAXBYTES+1 
  *  maximum bytes per character including NUL terminator
  */
-#ifdef UTF8_SUPPORT_FULL_UCS4
-# ifndef UTF8_MAXBYTES
-#  define UTF8_MAXBYTES  6
-# endif
-# ifndef UTF8_MAXBYTES1
-#  define UTF8_MAXBYTES1 7
-# endif
-#else
-# ifndef UTF8_MAXBYTES
-#  define UTF8_MAXBYTES  4
-# endif
-# ifndef UTF8_MAXBYTES1
-#  define UTF8_MAXBYTES1 5
-# endif
+#ifndef UTF8_MAXBYTES
+# define UTF8_MAXBYTES  4
+#endif
+#ifndef UTF8_MAXBYTES1
+# define UTF8_MAXBYTES1 5
 #endif
 /*--/moo--*/
 
 /* is c the start of a utf8 sequence? */
 #define isutf(c) (((c)&0xC0)!=0x80)
 
-/* convert UTF-8 data to wide character */
-int u8_toucs(u_int32_t *dest, int sz, char *src, int srcsz);
-
-/* the opposite conversion */
-int u8_toutf8(char *dest, int sz, u_int32_t *src, int srcsz);
-
 /* moo: get byte length of character number, or 0 if not supported */
 int u8_wc_nbytes(u_int32_t ch);
 
-/* moo: compute required storage for UTF-8 encoding of 's[0..n-1]' */
-int u8_wcs_nbytes(u_int32_t *ucs, int size);
-
 /* single character to UTF-8, no NUL termination */
 int u8_wc_toutf8(char *dest, u_int32_t ch);
 
@@ -67,22 +42,10 @@
 /* byte offset to character number */
 int u8_charnum(char *s, int offset);
 
-/* return next character, updating an index variable */
-u_int32_t u8_nextchar(char *s, int *i);
-
 /* move to next character */
 void u8_inc(char *s, int *i);
 
 /* move to previous character */
 void u8_dec(char *s, int *i);
 
-/* moo: move pointer to next character */
-void u8_inc_ptr(char **sp);
-
-/* moo: move pointer to previous character */
-void u8_dec_ptr(char **sp);
-
-/* returns length of next utf-8 sequence */
-int u8_seqlen(char *s);
-
 #endif /* S_UTF8_H */
